Documentation: Add tutorial about how to use partition mode on UP2 (#1547)
projectacrn/acrn-hypervisor: doc/tutorials/using_partition_mode_on_up2.rst

.. _partition_mode:

Using partition mode on UP2

> ===========================

Reviewer: ```suggestion
###########################
```

> This tutorial describes all the steps you need to take in order to use partition mode with the ACRN hypervisor on UP2.

Reviewer: Please keep line length < 80 characters when possible

Author: @dbkinder PR(https://github.com/projectacrn/acrn-hypervisor/pull/1558) has extended line length
to 120 instead of 80. Should we keep align with it.

> This tutorial describes all the steps you need to take in order to use partition mode with the ACRN hypervisor on UP2.

Reviewer: ```suggestion
This tutorial describes the steps needed to use partition mode with the ACRN hypervisor on UP2.
```

Enabling partition mode
***********************

> 1. Make the following changes in ``hypervisor/arch/x86/Kconfig`` :

Reviewer 2: This really should be done using `make menuconfig`, i.e. something like:
```
make defconfig PLATFORM=sbl
make menuconfig
```
and change both options

   .. code-block:: none

      diff --git a/hypervisor/arch/x86/Kconfig b/hypervisor/arch/x86/Kconfig
      index d4b41165..c279cd7f 100644
      --- a/hypervisor/arch/x86/Kconfig
      +++ b/hypervisor/arch/x86/Kconfig
      @@ -13,7 +13,7 @@ endchoice

      choice
          prompt "Hypervisor mode"
      -       default SHARING_MODE
      +       default PARTITION_MODE

      config SHARING_MODE
          bool "Sharing mode"
      @@ -72,7 +72,7 @@ endchoice
      config SERIAL_MMIO_BASE
          hex "Base address of serial MMIO region"
          depends on SERIAL_MMIO
      -       default 0xfc000000
      +       default <Your serial mmio addr>

> 2. Modify the ``vpci_vdev_arrary`` structure (one per vm) in ``hypervisor/partition/vm_description.c`` per 

Reviewer: Changing the ACRN source code doesn't seem right either?  What happens when a new release of ACRN is made?  Is there a board or platform configuration parameter when building ACRN?  Having the user change ACRN source code seems unsupportable?

Reviewer 2: I believe there is a series of patches coming about built-in VM configurations: https://lists.projectacrn.org/g/acrn-dev/message/13274. But until then, modifying the source code is the only option.

Can we all remember to modify this tutorial once the patch lands in the code?

Reviewer 2: At second thought, I think it's not related as this one uses `acrn-dm` which is **not** used in the partition mode. Is there a similar mechanism coming or is editing the source code the right option here?

Author: Directly changing the ACRN source code seems weird to me as well. But for current implementation, this is the only way we configure the assigned devices info and boot parameters.

We can make those devices info configurable through menuconfig. But there will be a lot of  effort to do so.

@gvancuts @dbkinder @jren1 @ailun258. What's your opinions?

>    hardware PCI info about devices passthroughed to vm.

Reviewer 2: ```suggestion
   hardware PCI info about devices assigned to the vm.
```

3. Configure ``.bootargs`` for vm kernel command linue. A default command line is configured by default.

   Here we just configure ``root=`` option to help kernel locate the right rootfs and keep the other options
   as default.

   .. code-block:: none

      @@ -171,7 +180,8 @@ struct vm_description_array vm_desc_partition = {
                                      .start_hpa = 0x100000000UL,
                                      .mem_size = 0x20000000UL, /* uses contiguous memory from host */
                                      .vm_vuart = true,
      -                               .bootargs = "root=/dev/sda rw rootwait noxsave maxcpus=2 nohpet console=hvc0 \
      +                               .bootargs = "root=PARTUUID=<Your rootfs PARTUUID> rw rootwait noxsave maxcpus=2 nohpet console=hvc0 \
                                                    console=ttyS0 no_timer_check ignore_loglevel log_buf_len=16M \
                                                    consoleblank=0 tsc=reliable xapic_phys",
                                      .vpci_vdev_array = &vpci_vdev_array1,

4. Build ACRN with the above changes.

   .. code-block:: none

      $ make PLATFORM=sbl

> 5. Install ubuntu on your UP2 and add one menuentry in ``<$efi_root>/grub/grub.cfg``

Reviewer 2: Can (should?) Ubuntu be installed on the eMMC?

   .. code-block:: none

      menuentry 'ACRN Partition Mode' --class ubuntu --class gnu-linux --class gnu --class os $menuentry_id_option 'gnulinux-simple-e23c76ae-b06d-4a6e-ad42-46b8eedfd7d3' {
              recordfail
              load_video
              gfxmode $linux_gfx_mode
              insmod gzio
              insmod part_gpt
              insmod ext2

              echo 'Loading partition mode hypervisor ...'
              multiboot /acrn.32.out
              module /bzImage
      }

>     Note: Make sure SATA and USB related modules are configured as "Y" of bzImage.

Reviewer 2: @dbkinder should we not use `.. note::` here?

> 6. Copy ``acrn.32.out`` and bzImage to ``<$efi_root>/``

Reviewer 2: Where does the `bzImage` kernel come from? We need instructions on  how to build it, or a pointer where one can download it.

Author: Sure, will do it in next version.

> 7. Insert Udisk or/and SATA SSD passthroughed to vms with rootfs installed into UP2 and then reboot board to 

Reviewer 2: What's the Udisk?

>    EFI shell. Then execute the following command to load acrn.

Reviewer: ```suggestion
   EFI shell. Then execute the following command to load the ACRN hypervisor:
```

   .. code-block:: none

>       $ EFI\ubuntu\grubx64.efi

Reviewer 2: We should put a complete example, using the EFI shell prompt as well (this one looks like a Linux terminal prompt). The series would be something like:
```
Shell> fs0:
fs0:\> EFI\ubuntu\grubx64.efi
```

With `fs0:` being potentially something different (like `fs1:`)

Author: Yes, it makes more sense. Will do it in next version.

> 

Reviewer 2: Add one step to instruct the user to select the `ACRN Partition Mode` menu entry

Author: Sure, will do.

> 8. Connect the serial port of the board and using the following commandto check if vms are boot up 

Reviewer: ```suggestion
8. Connect the serial port of the board and using the following command to check if vms booted 
```

> 8. Connect the serial port of the board and using the following commandto check if vms are boot up 

Reviewer: Connect the serial port to what?

> 8. Connect the serial port of the board and using the following commandto check if vms are boot up 

Reviewer 2: ```suggestion
8. Connect the serial port of the board and use the following command to check if vms booted
```
